libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,333 @@
 package bridge
 
 import (
-	"bytes"
 	"context"
 	"errors"
 	"fmt"
 	"net"
 
 	"github.com/containerd/log"
+	"github.com/docker/docker/libnetwork/iptables"
 	"github.com/docker/docker/libnetwork/netutils"
+	"github.com/docker/docker/libnetwork/portallocator"
+	"github.com/docker/docker/libnetwork/portmapper"
 	"github.com/docker/docker/libnetwork/types"
-	"github.com/ishidawataru/sctp"
 )
 
-func (n *bridgeNetwork) allocatePorts(ep *bridgeEndpoint, reqDefBindIP net.IP, ulPxyEnabled bool) ([]types.PortBinding, error) {
-	if ep.extConnConfig == nil || ep.extConnConfig.PortBindings == nil {
-		return nil, nil
-	}
+type portBinding struct {
+	*types.PortBinding
+	proxy portmapper.UserlandProxy
+}
 
-	defHostIP := net.IPv4zero // 0.0.0.0
-	if reqDefBindIP != nil {
-		defHostIP = reqDefBindIP
+// addPortMappings takes cfg, the configuration for port mappings, selects host
+// ports when ranges are given, starts docker-proxy or its dummy to reserve
+// host ports, and sets up iptables NAT/forwarding rules as necessary. If
+// anything goes wrong, it will undo any work it's done and return an error.
+// Otherwise, the returned slice of portBinding has an entry per address
+// family (if cfg describes a mapping for 'any' host address, it's expanded
+// into mappings for IPv4 and IPv6, because that's how the mapping is presented
+// in 'inspect'). HostPort and HostPortEnd in each returned portBinding are set
+// to the selected and reserved port.
+func (n *bridgeNetwork) addPortMappings(
+	epAddrV4, epAddrV6 *net.IPNet,
+	cfg []types.PortBinding,
+	defHostIP net.IP,
+) (_ []*portBinding, retErr error) {
+	if len(defHostIP) == 0 {
+		defHostIP = net.IPv4zero
+	} else if addr4 := defHostIP.To4(); addr4 != nil {
+		// Unmap the address if it's IPv4-mapped IPv6.
```

Comment:
Because iptables vs. ip6tables.

Reply:
ip(6)tables does not care either. An IPv4-mapped IPv6 address is a transitional mechanism **for userspace** to be able to write dual-stack apps using only IPv6 APIs. A socket bound to such an address will transmit and receive IPv4 packets over the wire, be filtered by iptables (not ip6tables), etc. Do like the `net` package does and treat 16-byte IPv4 addresses as semantically identical to 4-byte IPv4 addresses. Use the idiom `isIPv4 := addr.To4() != nil` everywhere instead of `isIPv4 := len(addr) == 4` and there is no need to "unmap" anything, or risk bugs if addresses weren't unmapped in some code path or another. I would go so far as to advise _against_ unmapping `net.IP` values here so that any downstream code which incorrectly checks the slice length to determine address family behaves incorrectly in the common case.

Reply:
An `iptables` rule with an IPv4-mapped IPv6 address will just-work. It's an IPv4 address.

An `ip6tables` rule with that address can be created, but it'll be useless because the packet it's aiming for will take the IPv4 route.

So, rules need to go into the right tables. The incorrect downstream behaviour you mention would be a bit invisible; the rule would be created normally, it just wouldn't do anything. I don't think that's something we want to invite.

There was one place where I compared lengths of addresses, knowing they were unmapped. It's gone now.